File: 프로그래머스/3/132266. 부대복귀/부대복귀.swift
import Foundation

func solution(_ n:Int, _ roads:[[Int]], _ sources:[Int], _ destination:Int) -> [Int] {
    var graph = Array(repeating: [Int](), count: n+1)
    for road in roads {
        graph[road[0]].append(road[1])
        graph[road[1]].append(road[0])
    }
    
    var distance = Array(repeating: -1, count: n+1)
    distance[destination] = 0

    var queue = [destination] //큐
    var index = 0
    
    while index < queue.count {
        let current = queue[index]
        index += 1
        
        for neighbor in graph[current] {
            if distance[neighbor] == -1 {
                distance[neighbor] = distance[current] + 1
                queue.append(neighbor)
            }
        }
    }

    return sources.map { distance[$0] }
}

File: 프로그래머스/3/132266. 부대복귀/부대복귀.py
from collections import deque

def solution(n, roads, sources, destination):
    graph = [[] for _ in range(n + 1)]
    for s, e in roads:
        graph[s].append(e)
        graph[e].append(s)
    
    # destination에서 출발하는 BFS!(역방향)
    distance = [-1] * (n + 1)
    distance[destination] = 0
    q = deque([destination])
    
    while q:
        now = q.popleft()
        for neighbor in graph[now]:
            if distance[neighbor] == -1:
                distance[neighbor] = distance[now] + 1
                q.append(neighbor)
    
    # 각 source에 대해 거리값을 찾아서 answer에 추가
    return [distance[s] for s in sources]
